### ISSUE TITLE
Supported Python and Sphinx versions

### ISSUE THREAD

**Issue by choldgraf** [sphinx-gallery/sphinx-gallery#404]
Recently in https://github.com/matplotlib/matplotlib/pull/11936 we had an interesting conversation about the difficulty of supporting multiple Sphinx and/or Python versions with Sphinx Gallery. I realized that I'm not exactly sure what is the Sphinx-Gallery team plan for this, and I wonder if it's worth having this conversation explicitly.

My 2 cents is this: Sphinx Gallery is probably not part of the "production pipeline" for research groups/companies/etc in the way that other analysis code is, and as such as can be a bit more aggressive about not supporting old versions of Python, Sphinx, etc. This could potentially help us reduce the maintenance burden and use more modern tools in Python. It might also help us attract more contributors, since I've seen a few times now people who don't contribute because they don't want to write Python 2.X code.

I'd love to hear people's thoughts on this!  (cc @anntzer for context, since you kinda sparked my interest in this issue :-) )

**Comment by GaelVaroquaux** [comment]
I don't particularly want to hang on to Python 3: it's currently being
dropped by most packages.

If we drop Python 3, we should do a release before dropping it, and
mention in the release notes.

**Comment by choldgraf** [comment]
Unless you are being particularly progressive, I am going to assume you mean python 2 😊

**Comment by Cadair** [comment]
+1 to dropping Python 2.

with regard to sphinx versions, I think something like last two versions (i.e 1.6 and 1.7) makes sense, as the sphinx version upgrade can be a little lumpy.

**Comment by choldgraf** [comment]
Here's a proposal: we freeze features and cut a new release relatively soon. Then after the release, begin testing only against the latest 2 major Sphinx versions, and latest 3 or 4 major versions of python 3.X

Thoughts? Thumbs up/down?

**Comment by GaelVaroquaux** [comment]
> Then after the release, begin testing only against the latest 2 major Sphinx
> versions, and latest 3 or 4 major versions of python 3.X

Latest 2 major Sphinx versions is a time window of 2 years (as seen from
https://pypi.org/project/Sphinx/#history). I find it a little short,
given that updating Sphinx a recurrent problem. Not an absolute blocker,
but we would be doing our users (downstream projects) a service to
maintain longer backward compatibility with sphinx.

**Comment by larsoner** [comment]
For packages used primarily by typical Python users I agree that 2 years is a bit short. It seems like SG has a different user base from most Python packages, though, as it's designed primarily to be used by Python developers and package maintainers. I expect far fewer use cases where someone needs it to work on an old CentOS or Ubuntu distribution, and in cases where it doesn't, the people are probably better equipped to be able to work around this. For me this trims the window that we should aim to support, so I'm fine with 2 sphinx 3-4 Python versions.

**Comment by choldgraf** [comment]
what @larsoner makes sense to me. Is there any way for us to determine how many python 2.x and 3.x users have installed via pip?

**Comment by larsoner** [comment]
No, not sure. I'd be okay with moving to Py3k-only after cutting the next release (soon).

**Comment by choldgraf** [comment]
OK, I've done a couple of things to follow-up on this issue, which I'll now close.

* Created milestones for [v0.2.1](https://github.com/sphinx-gallery/sphinx-gallery/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+milestone%3Av0.2.1) and [v0.3.0](https://github.com/sphinx-gallery/sphinx-gallery/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+milestone%3Av0.3.0)

Added the following more actionable issues for the two topics discussed here:

* #405 discusses dropping 2.X support
* #407 discusses dropping Sphinx versions >= 2/3 in the past

I've assigned each of these issues to the v0.3.0 release milestone (the one coming after v0.2.1). Gonna close this issue so that we can follow up separately in those!
